Create annotation links in batches if there are too many links

manics opened the pull request:
# What this PR does

#4775 added grouped-batch creation of MapAnnotations and links. This was required so that if a single MapAnnotation had multiple parents the MapAnnotation would not be incorrectly duplicated. However a small number of MapAnnotations have a very large number of parents which results in the Ice maximum message size being exceeded. This implements an additional method which creates the MapAnnotation and the links in separate steps if a batch size is exceeded.

# Testing this PR

Run `omero metadata populate` on a dataset which involves the creation of a `MapAnnotation` with an extremely large number of parents.

# Related reading

- https://trello.com/c/76es10Kg/23-rfe-populate-metadata-is-broken

manics committed:
Handle annotations with very large numbers of links

manics committed:
Change batch size integration tests: replace 1000 with 1

It makes more sense to test small batch sizes since the integration tests use very few annotations.
The alternative is to keep batch size 1000, but this unnecessarily increases the time to run the tests.

atarkowska commented:
I know that is not a part of that pr but would it be possible to trim keys and values? looks like http://10.0.51.193/webclient/?show=gene-SPTLC3%20 contains space in the end

manics commented:
Do you know where that's coming from? Is it a bug in the parser, or in the CSV?

manics commented:
OK, looks like the CSV:

```
idr-metadata (master)$ git grep 'SPTLC3 '
idr0009-simpson-secretion/screenA/idr0009-screenA-annotation.txt:0121-03--2006-02-10,121,215,I23,225087,CUUUAUUGUCCUCAUAGACtc,GUCUAUGAGGACAAUAAAGtt,ENSG00000172296,SPTLC3,,,pass,,SPTLC3 ,-1.07,173,3,,0.67,,,no,no,enhanced secretion,increased rate of protein secretion,CMPO_0000246,,,,,,
idr0009-simpson-secretion/screenA/idr0009-screenA-annotation.txt:0121-15--2006-01-28,121,215,I23,225087,CUUUAUUGUCCUCAUAGACtc,GUCUAUGAGGACAAUAAAGtt,ENSG00000172296,SPTLC3,,,pass,,SPTLC3 ,-1.07,173,3,,0.67,,,no,no,enhanced secretion,increased rate of protein secretion,CMPO_0000246,,,,,,
idr0009-simpson-secretion/screenA/idr0009-screenA-annotation.txt:0121-17--2006-05-04,121,215,I23,225087,CUUUAUUGUCCUCAUAGACtc,GUCUAUGAGGACAAUAAAGtt,ENSG00000172296,SPTLC3,,,pass,,SPTLC3 ,-1.07,173,3,,0.67,,,no,no,enhanced secretion,increased rate of protein secretion,CMPO_0000246,,,,,,
idr0009-simpson-secretion/screenA/idr0009-screenA-annotation.txt:0121-22--2006-03-11,121,215,I23,225087,CUUUAUUGUCCUCAUAGACtc,GUCUAUGAGGACAAUAAAGtt,ENSG00000172296,SPTLC3,,,pass,,SPTLC3 ,-1.07,173,3,,0.67,,,no,no,enhanced secretion,increased rate of protein secretion,CMPO_0000246,,,,,,
idr0009-simpson-secretion/screenA/idr0009-screenA-annotation.txt:0121-25--2006-05-22,121,215,I23,225087,CUUUAUUGUCCUCAUAGACtc,GUCUAUGAGGACAAUAAAGtt,ENSG00000172296,SPTLC3,,,pass,,SPTLC3 ,-1.07,173,3,,0.67,,,no,no,enhanced secretion,increased rate of protein secretion,CMPO_0000246,,,,,,
idr0009-simpson-secretion/screenA/idr0009-screenA-annotation.txt:0121-28--2006-03-06,121,215,I23,225087,CUUUAUUGUCCUCAUAGACtc,GUCUAUGAGGACAAUAAAGtt,ENSG00000172296,SPTLC3,,,pass,,SPTLC3 ,-1.07,173,3,,0.67,,,no,no,enhanced secretion,increased rate of protein secretion,CMPO_0000246,,,,,,
idr0009-simpson-secretion/screenA/idr0009-screenA-annotation.txt:0121-30--2006-02-20,121,215,I23,225087,CUUUAUUGUCCUCAUAGACtc,GUCUAUGAGGACAAUAAAGtt,ENSG00000172296,SPTLC3,,,pass,,SPTLC3 ,-1.07,173,3,,0.67,,,no,no,enhanced secretion,increased rate of protein secretion,CMPO_0000246,,,,,,
idr0009-simpson-secretion/screenA/idr0009-screenA-annotation.txt:0121-33--2006-05-14,121,215,I23,225087,CUUUAUUGUCCUCAUAGACtc,GUCUAUGAGGACAAUAAAGtt,ENSG00000172296,SPTLC3,,,pass,,SPTLC3 ,-1.07,173,3,,0.67,,,no,no,enhanced secretion,increased rate of protein secretion,CMPO_0000246,,,,,,
```

@eleanorwilliams should spaces be automatically trimmed from all CSV values?

manics commented:
Yet another alternative might be to strip the space in the bulk-map config file of the problematic studies using `{{ value | trim | urlencode }}` http://jinja.pocoo.org/docs/dev/templates/#trim

aleksandra-tarkowska commented:
happy for the second solution, will update config files

atarkowska commented:
ok tested that and looks like mitocheck was killed

```
ERROR:omero.util.populate_metadata:Missing primary keys (Continuing): Missing primary key fields: set(['Gene Symbol']) [('Gene Identifier', ['ENSG00000198014']), ('Gene Identifier URL', [u'http://www.ensembl.org/id/ENSG00000198014']), ('Gene Symbol', [])]
re-ann.bash: line 18:  2427 Killed                  omero metadata populate --context bulkmap --cfg $YML $OBJID
```

then tried the same by hand

```
ERROR:omero.util.populate_metadata:Missing primary keys (Continuing): Missing primary key fields: set(['Gene Symbol']) [('Gene Identifier', ['ENSG00000204511']), ('Gene Identifier URL', [u'http://www.ensembl.org/id/ENSG00000204511']), ('Gene Symbol', [])] 
ERROR:omero.util.populate_metadata:Missing primary keys (Continuing): Missing primary key fields: set(['Gene Symbol']) [('Gene Identifier', ['ENSG00000198014']), ('Gene Identifier URL', [u'http://www.ensembl.org/id/ENSG00000198014']), ('Gene Symbol', [])] 
Killed
```

@joshmoore mentioned that this is likely the OOMkiller.

```
[1951181.862927] Out of memory: Kill process 7500 (python) score 471 or sacrifice child
[1951181.862929] Killed process 7500 (python) total-vm:15987952kB, anon-rss:15527884kB, file-rss:0kB
```

joshmoore committed:
populate_metadata: reduce memory usage

BulkToMapAnnotation for large screens (+100K wells)
was being killed by the OOMKiller. Strategies include:

 * unload linked objects
 * drop `andReturn` where possible
 * return only sizes when possible
 * write actively in loops

joshmoore commented:
@manics : could you cherry-pick https://github.com/openmicroscopy/openmicroscopy/pull/4925/commits/29b3902317b63e2657e54bca9c1021301a23e9af here?

atarkowska commented:
> Yet another alternative might be to strip the space in the bulk-map config file of the problematic studies using {{ value | trim | urlencode }} http://jinja.pocoo.org/docs/dev/templates/#trim

Above requires adding `clientvalue: {{ value }}` on each single name in `*bulkmap-config.yml`, I think that is not what we aim? If we cannot automatically fix the script could I suggest fixing csv files?

manics commented:
@joshmoore Done

@aleksandra-tarkowska One of my ideas behind the `bulk-config.yml` setup was to avoid having to make code changes to do data cleanup or manipulation. I think fixing the CSV is the right way to go, since the CSV is meant to be a curated version of the raw data from submitters.

Since it's only  one screen changing `bulk-config.yml` for that screen might be faster though.

And longer term: I think having global `bulk-config.yml` files that can be included in the per-screen ones could be an option.

joshmoore committed:
populate_metadata: don't store plate data

Storing all of the plate data for idr0016 led to
death-by-OOM-killer during *initialization*. Now,
Ice objects are not being stored but wrapper very
thin Data objects.

joshmoore commented:
@manics : idr0016 done (4M annotations in approx 4 hours). Can you pick&push https://github.com/joshmoore/openmicroscopy/commit/dc2eb653ff78b8f8c67d8656f7a50be3abb8fb6f?

manics commented:
Done

joshmoore commented:
Merging for inclusion in 0.3.0-rc3

atarkowska commented:
--rebased-to #5227
